core/federated/federate.c

```diff
@@ -2355,97 +2335,101 @@ void handle_stop_granted_message(environment_t* env) {
+        lf_mutex_lock(&env[i].mutex);
+
+        // Sanity check.
+        if (lf_tag_compare(received_stop_tag, env[i].current_tag) <= 0) {
+            lf_print_error("RTI granted a MSG_TYPE_STOP_GRANTED tag that is equal to or less than this federate's current tag " PRINTF_TAG ". "
+                    "Stopping at the next microstep instead.",
+                    env[i].current_tag.time - start_time, env[i].current_tag.microstep);
+            received_stop_tag = env[i].current_tag;
+            received_stop_tag.microstep++;
+        }
+
+        env[i].stop_tag = received_stop_tag;
+        LF_PRINT_DEBUG("Setting the stop tag to " PRINTF_TAG ".",
+                    env[i].stop_tag.time - start_time,
+                    env[i].stop_tag.microstep);
+
+        _lf_decrement_tag_barrier_locked(&env[i]);
+        // We signal instead of broadcast under the assumption that only
+        // one worker thread can call wait_until at a given time because
+        // the call to wait_until is protected by a mutex lock
+        lf_cond_signal(&env[i].event_q_changed);
+        lf_mutex_unlock(&env[i].mutex);
+    }
 }
 
 /**
  * Handle a MSG_TYPE_STOP_REQUEST message from the RTI.
- *
- * This function assumes the caller does not hold
- * the mutex lock, therefore, it acquires it.
- * @param env The environment of the federate
  */
-void handle_stop_request_message(environment_t* env) {
-    assert(env != GLOBAL_ENVIRONMENT);
-
+void handle_stop_request_message() {
     size_t bytes_to_read = MSG_TYPE_STOP_REQUEST_LENGTH - 1;
     unsigned char buffer[bytes_to_read];
     read_from_socket_errexit(_fed.socket_TCP_RTI, bytes_to_read, buffer,
             "Failed to read stop request from RTI.");
-
-    // Acquire a mutex lock to ensure that this state does change while a
-    // message is being used to determine a TAG.
-    lf_mutex_lock(&env->mutex);
-    // Ignore the message if this federate originated a request.
-    // The federate is already blocked is awaiting a MSG_TYPE_STOP_GRANTED message.
-    if (_fed.sent_a_stop_request_to_rti == true) {
-        lf_mutex_unlock(&env->mutex);
-        return;
-    }
-
     tag_t tag_to_stop = extract_tag(buffer);
 
     // Trace the event when tracing is enabled
     tracepoint_federate_from_rti(_fed.trace, receive_STOP_REQ, _lf_my_fed_id, &tag_to_stop);
-
     LF_PRINT_LOG("Received from RTI a MSG_TYPE_STOP_REQUEST message with tag " PRINTF_TAG ".",
              tag_to_stop.time - start_time,
              tag_to_stop.microstep);
 
-    // Encode the current logical time plus one microstep
-    // or the requested tag_to_stop, whichever is bigger.
-    if (lf_tag_compare(tag_to_stop, env->current_tag) <= 0) {
-        // Can't stop at the requested tag. Make a counteroffer.
-        tag_to_stop = env->current_tag;
-        tag_to_stop.microstep++;
+    // If we have previously sent to the RTI a stop request
+    // or we have previously received from the RTI a stop request,
```

Comment:
As noted above. I think it would be clearer with two flags. One for having sent stop requests and one for having received stop requests